pkg/app/pipedv1/livestatereporter/livestatereporter.go

```diff
@@ -0,0 +1,171 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package livestatereporter provides a piped component
+// that reports the changes as well as full snapshot about live state of registered applications.
+package livestatereporter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipecd/pkg/app/server/service/pipedservice"
+	"github.com/pipe-cd/pipecd/pkg/model"
+	pluginapi "github.com/pipe-cd/pipecd/pkg/plugin/api/v1alpha1"
+	"github.com/pipe-cd/pipecd/pkg/plugin/api/v1alpha1/livestate"
+)
+
+type applicationLister interface {
+	ListByPluginName(name string) []*model.Application
+}
+
+type apiClient interface {
+	ReportApplicationLiveState(ctx context.Context, req *pipedservice.ReportApplicationLiveStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateResponse, error)
+	ReportApplicationSyncState(ctx context.Context, req *pipedservice.ReportApplicationSyncStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationSyncStateResponse, error)
+}
+
+// Reporter represents a component that reports the snapshot about live state of registered applications.
+type Reporter interface {
+	Run(ctx context.Context) error
+}
+
+type reporter struct {
+	reporters []pluginReporter
+	logger    *zap.Logger
+}
+
+// NewReporter creates a new reporter.
+func NewReporter(appLister applicationLister, apiClient apiClient, plugins map[string]pluginapi.PluginClient, logger *zap.Logger) Reporter {
+	rlogger := logger.Named("live-state-reporter")
+	r := &reporter{
+		reporters: make([]pluginReporter, 0, len(plugins)),
+		logger:    rlogger,
+	}
+
+	for name, p := range plugins {
+		r.reporters = append(r.reporters, pluginReporter{
+			pluginName:            name,
+			snapshotFlushInterval: time.Minute,
+			appLister:             appLister,
+			apiClient:             apiClient,
+			pluginClient:          p,
+			logger:                rlogger.With(zap.String("plugin-name", name)),
+		})
+	}
+	return r
+}
+
+func (r *reporter) Run(ctx context.Context) error {
+	group, ctx := errgroup.WithContext(ctx)
+
+	for i, reporter := range r.reporters {
+		// Avoid starting all reporters at the same time to reduce the API call burst.
+		time.Sleep(time.Duration(i) * 10 * time.Second)
+		r.logger.Info("starting app live state reporter for plugin")
+
+		group.Go(func() error {
+			return reporter.Run(ctx)
+		})
```

Comment:
If there are N plugins, this code sleeps for `N(N+1) * 10s`. It's a bit long and the expected duration seems `N * 10s`. One more, the info log should contain the plugin name. So, I suggest 2 ways below.



1
```suggestion
		// Avoid starting all reporters at the same time to reduce the API call burst.
		group.Go(func() error {
			time.Sleep(time.Duration(i) * 10 * time.Second)
			reporter.logger.Info("starting app live state reporter for plugin")))

			return reporter.Run(ctx)
		})
```

2
```suggestion
		// Avoid starting all reporters at the same time to reduce the API call burst.
		time.Sleep(10 * time.Second)
		reporter.logger.Info("starting app live state reporter for plugin")

		group.Go(func() error {
			return reporter.Run(ctx)
		})
```

Reply:
@Warashi Thanks! I fixed

- wait 10s for starting every plugin's reporter [9b697dd](https://github.com/pipe-cd/pipecd/pull/5452/commits/9b697dd175e39c8bfac67336f44c0234bcdf49e1)
- Use logger of plugin reporter [24fb566](https://github.com/pipe-cd/pipecd/pull/5452/commits/24fb566fd8e4b82eb1c4e9a254adaf1153d02e08)

Also, I added the info log to notice starting main reporter. [a7217fc](https://github.com/pipe-cd/pipecd/pull/5452/commits/a7217fc818d6312215f12cd3fcce1bb2ac3015cd)